File: android/app/src/main/kotlin/com/khaledsamir/reelseek/viewmodel/DetailViewModel.kt
package com.khaledsamir.reelseek.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.khaledsamir.reelseek.model.Availability
import com.khaledsamir.reelseek.model.MediaType
import com.khaledsamir.reelseek.model.MovieDetails
import com.khaledsamir.reelseek.network.ReelseekApi
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

// Mirrors ios DetailViewModel, plus website behavior: switching country
// re-fetches availability only.
class DetailViewModel(private val api: ReelseekApi) : ViewModel() {
    var details by mutableStateOf<MovieDetails?>(null)
        private set
    var availability by mutableStateOf<Availability?>(null)
        private set
    var isLoading by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    private var loadedKey: Pair<Int, String>? = null

    fun load(tmdbId: Int, mediaType: MediaType, country: String) {
        val key = tmdbId to country
        if (loadedKey == key) return
        val sameTitle = loadedKey?.first == tmdbId
        loadedKey = key
        if (sameTitle) {
            // Country switch: details are unchanged, only availability moves.
            reloadAvailability(tmdbId, mediaType, country)
            return
        }
        viewModelScope.launch {
            isLoading = true
            try {
                val detailsDeferred = async { api.details(tmdbId, mediaType.key) }
                val availabilityDeferred = async {
                    runCatching { api.availability(tmdbId, mediaType.key, country) }.getOrNull()
                }
                details = detailsDeferred.await()
                availability = availabilityDeferred.await()
                errorMessage = null
            } catch (e: Exception) {
                errorMessage = e.message ?: "Failed to load details."
            } finally {
                isLoading = false
            }
        }
    }

    fun reloadAvailability(tmdbId: Int, mediaType: MediaType, country: String) {
        viewModelScope.launch {
            availability = runCatching {
                api.availability(tmdbId, mediaType.key, country)
            }.getOrNull()
        }
    }
}

File: ios/Reelseek/ViewModels/DetailViewModel.swift
import Foundation
import Observation

@Observable
@MainActor
final class DetailViewModel {
    var details: MovieDetails?
    var availability: Availability?
    var isLoading = false
    var errorMessage: String?

    func load(tmdbId: Int, mediaType: MediaType, country: String) async {
        isLoading = true
        defer { isLoading = false }
        do {
            async let d = MovieAPI.details(tmdbId: tmdbId, mediaType: mediaType)
            async let a = try? await MovieAPI.availability(
                tmdbId: tmdbId,
                mediaType: mediaType,
                country: country
            )
            self.details = try await d
            self.availability = await a
            self.errorMessage = nil
        } catch {
            self.errorMessage = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
        }
    }

    /// Country switch: details are unchanged, only availability moves.
    func reloadAvailability(tmdbId: Int, mediaType: MediaType, country: String) async {
        availability = try? await MovieAPI.availability(
            tmdbId: tmdbId,
            mediaType: mediaType,
            country: country
        )
    }
}
